File: ReactNative/IMSDK/react-native-tim-js/ios/manager/signalingManager.swift
import Foundation
import ImSDK_Plus

class SignalingManager {
	public func invite(param: [String: Any], resolve: @escaping RCTPromiseResolveBlock) {
        let v2TIMOfflinePushInfo = CommonUtils.getV2TIMOfflinePushInfo(param: param)

        if var invitee = param["invitee"] as? String,
            let data = param["data"] as? String,
            let timeout = param["timeout"] as? Int32,
            let onlineUserOnly = param["onlineUserOnly"] as? Bool {
            var inviteID = ""
            //    invitee为""crash，" "可以走到fail
            invitee = (invitee == "") ? " " : invitee
            inviteID = V2TIMManager.sharedInstance().invite(invitee, data: data, onlineUserOnly: onlineUserOnly, offlinePushInfo: v2TIMOfflinePushInfo, timeout: timeout, succ: {
                () -> Void in
                CommonUtils.resultSuccess(method: "invite", resolve: resolve, data: inviteID)
            }, fail: TencentImUtils.returnErrorClosures(method: "invite", resolve: resolve))
        }
	}

	public func inviteInGroup(param: [String: Any], resolve: @escaping RCTPromiseResolveBlock) {
        if let groupID = param["groupID"] as? String,
            let inviteeList = param["inviteeList"] as? Array<String>,
            let data = param["data"] as? String,
            let timeout = param["timeout"] as? Int32,
            let onlineUserOnly = param["onlineUserOnly"] as? Bool {
            var inviteID = ""
            //    inviteeList为空数组crash，非空可以走到fail
            if inviteeList.count == 0 {
                return CommonUtils.resultFailed(desc: "InviteeList can not be empty", code: -1, method: "inviteInGroup", resolve: resolve)
            }
            inviteID = V2TIMManager.sharedInstance().invite(inGroup: groupID, inviteeList: inviteeList, data: data, onlineUserOnly: onlineUserOnly, timeout: timeout, succ: {
                () -> Void in
                CommonUtils.resultSuccess(method: "inviteInGroup", resolve: resolve, data: inviteID)
            }, fail: TencentImUtils.returnErrorClosures(method: "inviteInGroup", resolve: resolve))
        }
	}

	public func cancel(param: [String: Any], resolve: @escaping RCTPromiseResolveBlock) {
        if let inviteID = param["inviteID"] as? String,
            let data = param["data"] as? String {

            V2TIMManager.sharedInstance().cancel(inviteID, data: data, succ: {
                () -> Void in
                CommonUtils.resultSuccess(method: "cancel", resolve: resolve, data: "ok");
            }, fail: TencentImUtils.returnErrorClosures(method: "cancel", resolve: resolve))
        }
	}

	public func accept(param: [String: Any], resolve: @escaping RCTPromiseResolveBlock) {
        if let inviteID = param["inviteID"] as? String,
            let data = param["data"] as? String {
                
            V2TIMManager.sharedInstance().accept(inviteID, data: data, succ: {
                () -> Void in
                CommonUtils.resultSuccess(method: "accept", resolve: resolve, data: "ok");
            }, fail: TencentImUtils.returnErrorClosures(method: "accept", resolve: resolve))
        }
	}

	public func reject(param: [String: Any], resolve: @escaping RCTPromiseResolveBlock) {
        if let inviteID = param["inviteID"] as? String,
            let data = param["data"] as? String {
                
            V2TIMManager.sharedInstance().reject(inviteID, data: data, succ: {
                () -> Void in
                CommonUtils.resultSuccess(method: "reject", resolve: resolve, data: "ok");
            }, fail: TencentImUtils.returnErrorClosures(method: "reject", resolve: resolve))
        }
	}

    // TODO 底层ios的名字是getSignallingInfo, 因此我调用的时候是两个ll
	public func getSignalingInfo(param: [String: Any], resolve: @escaping RCTPromiseResolveBlock) {
        if let messageID = param["msgID"] as? String {
            V2TIMManager.sharedInstance().findMessages([messageID], succ: {
                msgs in
                if(msgs!.isEmpty) {
                    CommonUtils.resultSuccess(method: "getSignalingInfo", resolve: resolve, data: "messages no found");
                }else{
                    let signalInfo = V2TIMManager.sharedInstance().getSignallingInfo(msgs![0]);
                    
                    var resultMap = [String:Any]();

                    if(signalInfo != nil){
                        resultMap = V2TIMSignalingInfoEntiry.getDict(info: signalInfo!);

                        CommonUtils.resultSuccess(method: "getSignalingInfo", resolve: resolve, data: resultMap)
                    }else {
                        CommonUtils.resultSuccess(method: "getSignalingInfo", resolve: resolve, data: signalInfo as Any)
                        
                    }
                    
                }
            }, fail:  TencentImUtils.returnErrorClosures(method: "getSignalingInfo", resolve: resolve))
        };
	}
	
	public func addInvitedSignaling(param: [String: Any], resolve: @escaping RCTPromiseResolveBlock) {
        let info = param["info"] as! [String: Any];
        let param = V2TIMSignalingInfo();
        if(info["inviteID"] != nil) {  // 邀请 ID
            param.inviteID = info["inviteID"] as? String;
        }
        if(info["groupID"] != nil) {     // 发起邀请所在群组
            param.groupID = info["groupID"] as? String;
        }
        if(info["inviteeList"] != nil) {    // 被邀请人列表
         
            param.inviteeList = info["inviteeList"] as? NSMutableArray;
        }
        if(info["data"] != nil) { // 
            param.data = info["data"] as? String;
        }
        if(info["timeout"] != nil) {
            param.timeout = info["timeout"] as! UInt32;
        }
        if(info["actionType"] != nil) {
            param.actionType = SignalingActionType(rawValue:info["actionType"] as! Int)!;
        }
        // 注意：ios不需要businessID
        V2TIMManager.sharedInstance().addInvitedSignaling(param, succ: {
            CommonUtils.resultSuccess(method: "addInvitedSignaling", resolve: resolve, data: "ok");
        }, fail: TencentImUtils.returnErrorClosures(method: "addInvitedSignaling", resolve: resolve))
	}


}

File: ReactNative/IMSDK/react-native-tim-js/android/src/main/java/com/reactnativetimjs/manager/SignalingManager.java
package com.reactnativetimjs.manager;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableMap;

import com.reactnativetimjs.util.CommonUtils;
import com.tencent.imsdk.v2.V2TIMCallback;
import com.tencent.imsdk.v2.V2TIMManager;
import com.tencent.imsdk.v2.V2TIMMessage;
import com.tencent.imsdk.v2.V2TIMOfflinePushInfo;
import com.tencent.imsdk.v2.V2TIMSignalingInfo;
import com.tencent.imsdk.v2.V2TIMSignalingListener;
import com.tencent.imsdk.v2.V2TIMValueCallback;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

public class SignalingManager {
    private static V2TIMSignalingListener signallistener;

    private static HashMap<String, String> invitedIDMap = new HashMap();
    private static HashMap<String, V2TIMSignalingListener> signalListenerList = new HashMap<>();

    public void addSignalingListener(Promise promise, ReadableMap arguments) {
        signallistener = new V2TIMSignalingListener() {
            @Override
            public void onReceiveNewInvitation(String inviteID, String inviter, String groupID,
                    List<String> inviteeList, String data) {
                HashMap<String, Object> res = new HashMap<String, Object>();
                res.put("inviteID", inviteID);
                res.put("inviter", inviter);
                res.put("groupID", groupID);
                res.put("inviteeList", inviteeList);
                res.put("data", data);
                makeSignalingListenerEventData("onReceiveNewInvitation", res);
            }

            @Override
            public void onInviteeAccepted(String inviteID, String invitee, String data) {
                HashMap<String, Object> res = new HashMap<String, Object>();
                res.put("inviteID", inviteID);
                res.put("invitee", invitee);
                res.put("data", data);
                makeSignalingListenerEventData("onInviteeAccepted", res);
            }

            @Override
            public void onInviteeRejected(String inviteID, String invitee, String data) {
                HashMap<String, Object> res = new HashMap<String, Object>();
                res.put("inviteID", inviteID);
                res.put("invitee", invitee);
                res.put("data", data);
                makeSignalingListenerEventData("onInviteeRejected", res);
            }

            @Override
            public void onInvitationCancelled(String inviteID, String inviter, String data) {
                HashMap<String, Object> res = new HashMap<String, Object>();
                res.put("inviteID", inviteID);
                res.put("inviter", inviter);
                res.put("data", data);
                makeSignalingListenerEventData("onInvitationCancelled", res);
            }

            @Override
            public void onInvitationTimeout(String inviteID, List<String> inviteeList) {
                HashMap<String, Object> res = new HashMap<String, Object>();
                res.put("inviteID", inviteID);
                res.put("inviteeList", inviteeList);
                makeSignalingListenerEventData("onInvitationTimeout", res);
            }
        };
        V2TIMManager.getSignalingManager().addSignalingListener(signallistener);
        CommonUtils.returnSuccess(promise, "add signaling listener success");
    }

    public void removeSignalingListener(Promise promise, ReadableMap arguments) {
        V2TIMManager.getSignalingManager().removeSignalingListener(signallistener);
        CommonUtils.returnSuccess(promise, "removeSignalingListener is done");
    }

    private <T> void makeSignalingListenerEventData(String eventType, T data) {
        CommonUtils.emmitEvent("signalingListener", eventType, data);
    }

    public void invite(Promise promise, ReadableMap arguments) {
        String invitee = arguments.getString("invitee");
        String data = arguments.getString("data");
        int timeout = arguments.getInt("timeout");
        boolean onlineUserOnly = arguments.getBoolean("onlineUserOnly");
        V2TIMOfflinePushInfo offlinePushInfo = new V2TIMOfflinePushInfo();
        HashMap<String, Object> offlinePushInfoParams = CommonUtils
                .convertReadableMapToHashMap(arguments.getMap("offlinePushInfo"));
        if (offlinePushInfoParams != null) {
            String title = (String) offlinePushInfoParams.get("title");
            String desc = (String) offlinePushInfoParams.get("desc");
            Boolean disablePush = (Boolean) offlinePushInfoParams.get("disablePush");
            String ext = (String) offlinePushInfoParams.get("ext");
            String iOSSound = (String) offlinePushInfoParams.get("iOSSound");
            Boolean ignoreIOSBadge = (Boolean) offlinePushInfoParams.get("ignoreIOSBadge");
            String androidOPPOChannelID = (String) offlinePushInfoParams.get("androidOPPOChannelID");
            if (title != null) {
                offlinePushInfo.setTitle(title);
            }
            if (desc != null) {
                offlinePushInfo.setDesc(desc);
            }
            if (offlinePushInfoParams.get("disablePush") != null) {
                offlinePushInfo.disablePush(disablePush);
            }
            if (ext != null) {
                offlinePushInfo.setExt(ext.getBytes());
            }
            if (iOSSound != null) {
                offlinePushInfo.setIOSSound(iOSSound);
            }
            if (offlinePushInfoParams.get("ignoreIOSBadge") != null) {
                offlinePushInfo.setIgnoreIOSBadge(ignoreIOSBadge);
            }
            if (androidOPPOChannelID != null) {
                offlinePushInfo.setAndroidOPPOChannelID(androidOPPOChannelID);
            }
        }
        final String current = String.valueOf(System.nanoTime());
        String id = V2TIMManager.getSignalingManager().invite(invitee, data, onlineUserOnly, offlinePushInfo,
                timeout, new V2TIMCallback() {

                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                        invitedIDMap.remove(current);
                    }

                    @Override
                    public void onSuccess() {
                        // Here, the id returned synchronously above is used
                        CommonUtils.returnSuccess(promise, invitedIDMap.get(current));
                        invitedIDMap.remove(current);
                    }
                });

        invitedIDMap.put(current, id);
    }

    public void inviteInGroup(Promise promise, ReadableMap arguments) {
        String groupID = arguments.getString("groupID");
        List<String> inviteeList = CommonUtils.convertReadableArrayToListString(arguments.getArray("inviteeList"));
        String data = arguments.getString("data");
        int timeout = arguments.getInt("timeout");
        boolean onlineUserOnly = arguments.getBoolean("onlineUserOnly");
        final String current = String.valueOf(System.nanoTime());
        String id = V2TIMManager.getSignalingManager().inviteInGroup(groupID, inviteeList, data, onlineUserOnly,
                timeout, new V2TIMCallback() {
                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                        invitedIDMap.remove(current);
                    }

                    @Override
                    public void onSuccess() {
                        CommonUtils.returnSuccess(promise, invitedIDMap.get(current));
                        invitedIDMap.remove(current);
                    }
                });
        invitedIDMap.put(current, id);

    }

    public void cancel(Promise promise, ReadableMap arguments) {
        String inviteID = arguments.getString("inviteID");
        String data = arguments.getString("data");
        V2TIMManager.getSignalingManager().cancel(inviteID, data, new V2TIMCallback() {
            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }

            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }
        });
    }

    public void accept(Promise promise, ReadableMap arguments) {
        String inviteID = arguments.getString("inviteID");
        String data = arguments.getString("data");
        V2TIMManager.getSignalingManager().accept(inviteID, data, new V2TIMCallback() {
            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }

            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }
        });
    }

    public void reject(Promise promise, ReadableMap arguments) {
        String inviteID = arguments.getString("inviteID");
        String data = arguments.getString("data");
        V2TIMManager.getSignalingManager().reject(inviteID, data, new V2TIMCallback() {
            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }

            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }
        });
    }

    public void getSignalingInfo(final Promise promise, ReadableMap arguments) {
        String msgID = arguments.getString("msgID");
        List<String> messageIdList = new LinkedList<>();
        messageIdList.add(msgID);
        V2TIMManager.getMessageManager().findMessages(messageIdList,
                new V2TIMValueCallback<List<V2TIMMessage>>() {
                    @Override
                    public void onSuccess(List<V2TIMMessage> v2TIMMessages) {
                        if (v2TIMMessages.size() != 1) {
                            CommonUtils.returnError(promise, -1, "message not found");
                        } else {
                            V2TIMSignalingInfo singnalInfo = V2TIMManager.getSignalingManager()
                                    .getSignalingInfo(v2TIMMessages.get(0));
                            if (singnalInfo == null) {
                                CommonUtils.returnSuccess(promise, null);
                                return;
                            }
                            CommonUtils.returnSuccess(promise,
                                    CommonUtils.convertV2TIMSignalingInfoToMap(singnalInfo));
                        }
                    }

                    @Override
                    public void onError(int code, String desc) {
                        CommonUtils.returnError(promise, code, desc);
                    }
                });

    }

    public void addInvitedSignaling(final Promise promise, ReadableMap arguments) {
        HashMap<String, Object> info = CommonUtils.convertReadableMapToHashMap(arguments.getMap("info"));
        V2TIMSignalingInfo param = new V2TIMSignalingInfo();
        if (info.get("inviteID") != null) {
            param.setInviteID((String) info.get("inviteID"));
        }
        if (info.get("groupID") != null) {
            param.setGroupID((String) info.get("groupID"));
        }
        if (info.get("inviter") != null) {
            param.setInviter((String) info.get("inviter"));
        }
        if (info.get("inviteeList") != null) {
            param.setInviteeList((List<String>) info.get("inviteeList"));
        }
        if (info.get("data") != null) {
            param.setData((String) info.get("data"));
        }
        if (info.get("timeout") != null) {
            param.setTimeout((Integer) info.get("timeout"));
        }
        if (info.get("actionType") != null) {
            param.setActionType((Integer) info.get("actionType"));
        }
        if (info.get("businessID") != null) {
            param.setBusinessID((Integer) info.get("businessID"));
        }
        V2TIMManager.getSignalingManager().addInvitedSignaling(param, new V2TIMCallback() {
            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtils.returnError(promise, code, desc);
            }
        });
    }
}
